design/baremetal-operator/image-builder-integration.md

```diff
@@ -0,0 +1,215 @@
+image.
+
+### Goals
+
+- Allow each host to use its own deploy ramdisk image customised with an
+  appropriate network configuration.
+
+### Non-Goals
+
+- Allow users who expect to provision a host to specify their own custom image
+  containing IPA.
+- Use Ironic with images that don't contain IPA.
+- Implement custom provisioners not based on Ironic.
+
+## Proposal
+
+### User Stories
+
+As a data center operator, I would like to manage hosts with metal³ and deploy
+them via virtualmedia on networks where there is no DHCP-like configuration
+service.
+
+## Design Details
+
+A new Custom Resource type, `RamDiskImageRequest` will be created in the
+metal3.io group. The Spec will contain the following fields:
+
+- Name of metaData Secret
+- Name of networkData Secret
+- Architecture
+
+The Status will contain the following fields:
+
+- Image URL
+- Name + Version of metaData Secret
+- Name + Version of networkData Secret
+- Error message
+- Ready condition
+- Error condition
+
+A default controller for this resource type will be implemented that simply
+sets a fixed URL (supplied in its own environment or command line) as the
+output. However, provision will be made for interested parties to implement
+alternate controllers to allow image customisation.
+
+A command-line switch (defaulting to off) will enable integration between the
+baremetal-operator and the image request controller.
+
+### Implementation Details/Notes/Constraints
+
+When integration is enabled, the controller portion of the baremetal-operator
+will update the `RamDiskImageRequest` with the same name and namespace as the
+`BareMetalHost`, creating it if it doesn't exist, to point to the metaData and
+networkData Secrets referenced in the `BareMetalHost` Spec. The Architecture
+will be obtained from the Host's hardware profile.
+
+In the Registration state, the ironic provisioner will avoid attempting to
+Manage the Node in Ironic until the Ready condition is true in the
+corresponding `RamDiskImageRequest` Status. It will then tell configure the
+Node to use the ramdisk Image from the URL in the status.
+
+The contents of the metaData and networkData Secrets are generally assumed to
+be in the same format output by OpenStack and consumable by cloud-init.
+However, for Metal³'s purposes this is not a requirement as the data is passed
+through directly. Authors of image customisation controllers are free to
+interpret the data in any way they see fit.
+
+The `BareMetalHost` should be set as the Owner of the `RamDiskImageRequest` so
+that the request gets deleted when the Host is deleted. This allows controllers
+to use the resource to perform cache management.
+
+Once the image URL is set, it shouldn't be changed. If the user wishes to
+rebuild the ramdisk with newer settings, they can do so by deleting the image
+request. The baremetal-operator should watch the image request and recreate it
+when this occurs.
```

Comment:
What happens to the provisioned node? What happens if the node is in flight?

Reply:
We'll update the settings in the node if they don't match, and pause during registration if the image is not available during registration. The next time ironic boots into the deploy image it'll use the new one.